lib/web_ui/lib/src/engine/surface/path_metrics.dart

```diff
@@ -524,10 +723,275 @@ class _EllipseSegmentResult {
   _EllipseSegmentResult();
 }
 
+// Given a vector dx, dy representing slope, normalize and return as [ui.Offset].
+ui.Offset _normalizeSlope(double dx, double dy) {
+  final double length = math.sqrt(dx * dx + dy * dy);
+  return length < kEpsilon
+      ? ui.Offset(0.0, 0.0)
+      : ui.Offset(dx / length, dy / length);
+}
+
+class _SurfaceTangent extends ui.Tangent {
+  const _SurfaceTangent(ui.Offset position, ui.Offset vector, this.t)
+      : assert(position != null),
+        assert(vector != null),
+        assert(t != null),
+        super(position, vector);
+
+  // Normalized distance of tangent point from start of a contour.
+  final double t;
+}
+
 class _PathSegment {
   _PathSegment(this.segmentType, this.distance, this.points);
 
   final int segmentType;
   final double distance;
   final List<double> points;
+
+  _SurfaceTangent computeTangent(double t) {
+    switch (segmentType) {
+      case PathCommandTypes.lineTo:
+        // Simple line. Position is simple interpolation from start to end point.
+        final double xAtDistance = (points[2] * t) + (points[0] * (1.0 - t));
+        final double yAtDistance = (points[3] * t) + (points[1] * (1.0 - t));
+        return _SurfaceTangent(ui.Offset(xAtDistance, yAtDistance),
+            _normalizeSlope(points[2] - points[0], points[3] - points[1]), t);
+      case PathCommandTypes.bezierCurveTo:
+        return tangentForCubicAt(t, points[0], points[1], points[2], points[3],
+            points[4], points[5], points[6], points[7]);
+      case PathCommandTypes.quadraticCurveTo:
+        return tangentForQuadAt(t, points[0], points[1], points[2], points[3],
+            points[4], points[5]);
+      default:
+        throw UnsupportedError('Invalid segment type');
+    }
+  }
+
+  _SurfaceTangent tangentForQuadAt(double t, double x0, double y0, double x1,
+      double y1, double x2, double y2) {
+    assert(t >= 0 && t <= 1);
+    final _SkQuadCoefficients _quadEval =
+        _SkQuadCoefficients(x0, y0, x1, y1, x2, y2);
+    final ui.Offset pos = ui.Offset(_quadEval.evalX(t), _quadEval.evalY(t));
+    // Derivative of quad curve is 2(b - a + (a - 2b + c)t).
+    // If control point is at start or end point, this yields 0 for t = 0 and
+    // t = 1. In that case use the quad end points to compute tangent instead
+    // of derivative.
+    final ui.Offset tangentVector = ((t == 0 && x0 == x1 && y0 == y1) ||
```

Comment:
Based on comment above, shouldn't we also be checking for `t == 1` here?

Reply:
thx for spotting incorrect t == 0. Corrected.